### ISSUE TITLE
Wrap padding

### ISSUE THREAD

**Issue by gcskoenig** [pytorch/pytorch#3858]
Hi all,

I could not find any padding functionality that equals the "wrap" padding from numpy.

E.g. this:
[[0 1 2]
 [3 4 5]
 [6 7 8]]

Shall be padded like this:
[[8 6 7 8 6]
 [2 0 1 2 0]
 [5 3 4 5 3]
 [8 6 7 8 6]
 [2 0 1 2 0]]

Would be of great value!

**Comment by zou3519** [comment]
I think this is what you want:
```
import torch
n = 3  # n by n matrix
x = torch.Tensor([[0, 1, 2],[3, 4, 5],[6, 7, 8]])
x.repeat(n, n)[(n-1):(2*n+1), (n-1):(2*n+1)]
```
Idea: repeat the tensor in a 3x3 grid, then crop out the part you need.

**Comment by harpone** [comment]
Here's a simple and pretty fast implementation for `[H, W]` shaped arrays:

```
def pad_circular(x, pad):
    """

    :param x: shape [H, W]
    :param pad: int >= 0
    :return:
    """
    x = torch.cat([x, x[0:pad]], dim=0)
    x = torch.cat([x, x[:, 0:pad]], dim=1)
    x = torch.cat([x[-2 * pad:-pad], x], dim=0)
    x = torch.cat([x[:, -2 * pad:-pad], x], dim=1)

    return x
```

For a `1000x1000` array I get this:
```
N = 1000

x = torch.randn(N, N).cuda()

%timeit pad_circular(x, 1)

> 38 µs ± 1.53 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

For the `repeat` way above I get CUDA OOM error...

**Comment by harpone** [comment]
@Douphoton it's just slicing and concatenation so it's differentiable.

BTW I made a mistake in the repeat implementation, now I got 263 µs ± 853 ns, so the "manual" version above is not *that* much faster...

**Comment by huangbiubiu** [comment]
A more general version inspired by @harpone To apply padding on a 3-d image (not padding on the channel dimension):


```python3

def pad_circular_nd(x: torch.Tensor, pad: int, dim) -> torch.Tensor:
    """
    :param x: shape [H, W]
    :param pad: int >= 0
    :param dim: the dimension over which the tensors are padded
    :return:
    """

    if isinstance(dim, int):
        dim = [dim]

    for d in dim:
        if d >= len(x.shape):
            raise IndexError(f"dim {d} out of range")

        idx = tuple(slice(0, None if s != d else pad, 1) for s in range(len(x.shape)))
        x = torch.cat([x, x[idx]], dim=d)

        idx = tuple(slice(None if s != d else -2 * pad, None if s != d else -pad, 1) for s in range(len(x.shape)))
        x = torch.cat([x[idx], x], dim=d)
        pass

    return x
```

Performance: similar to original implementation, expectedly.

```python3
N = 1000

x = torch.randn(N, N).cuda()

%timeit pad_circular(x, 1)
%timeit pad_circular_nd(x, 1, dim=[0,1])

# 1000 loops, best of 3: 455 µs per loop
# 1000 loops, best of 3: 509 µs per loop
```

**Comment by Chillee** [comment]
Closing due to https://github.com/pytorch/pytorch/pull/17240
